### PR TITLE
Fix bool mask IoU comparison for Detectron2 outputs

### LINKED ISSUE #127073: [inductor][cpu]detectron2_maskrcnn_r_101_fpn detectron2_maskrcnn_r_50_c4 accuracy check failed when freezing flag is on.
### 🐛 Describe the bug

This issue is detected in CI for PR: [#124715](https://github.com/pytorch/pytorch/pull/124715), which only occurs when freezing flag on.

Error Log:  [CI [Log]](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715)
[1467](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715#step:19:1468)WARNING:common:fp64 golden ref were not generated for detectron2_maskrcnn_r_101_fpn. Setting accuracy check to cosine
[1468](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715#step:19:1469)E0523 06:57:26.283015 139961353447232 torch/_dynamo/utils.py:1339] Accuracy failed: uint8 tensor did not match
[1469](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715#step:19:1470)E0523 06:57:26.283288 139961353447232 torch/_dynamo/utils.py:1306] Accuracy failed for key name pred_masks
[1470](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715#step:19:1471)E0523 06:57:26.283616 139961353447232 torch/_dynamo/utils.py:1306] Accuracy failed for key name instances

### Versions

 [CI [Log]](https://github.com/pytorch/pytorch/actions/runs/9202905546/job/25313957379?pr=124715)

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186441

CPU Inductor freezing can produce tiny numeric differences in Detectron2 mask logits. Those differences are thresholded into boolean pred_masks, so exact bool equality is too strict for these segmentation outputs. The benchmark harness already has an IoU-based bool mask comparator for this class of model, but torch._dynamo.utils.same() dropped use_iou_for_bool and iou_threshold when recursing through Detectron2 Instances objects. As a result, Instances.pred_masks still used exact equality and the affected Mask R-CNN models remained fail_accuracy under freezing.

Propagate the bool-mask IoU options through the object and numpy recursive same() paths, enable the existing IoU comparator for the two affected Detectron2 Mask R-CNN models, and update the CPU freezing expected-accuracy files to pass.

Fixes #127073

Generated by my agent

Test Plan:

python test/dynamo/test_utils.py -k TestUtils.test_same_iou_for_bool_propagates_to_instances

python -m unittest benchmarks.dynamo.test.TestDynamoBenchmark.test_detectron2_maskrcnn_uses_iou_for_bool_masks

Local Detectron2 benchmark workaround for detectron2_maskrcnn_r_101_fpn and detectron2_maskrcnn_r_50_c4 with --accuracy --inference --device cpu --inductor --freezing reported pass

git diff --check

lintrunner -a (fails on unrelated existing Pyrefly error in torch/autograd/graph.py for torch._C._remove_obj_from_tls)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says CI failed because the dynamic CPU max-autotune AMP freezing TorchBench lane reported speech_transformer as an accuracy improvement: pass while the expected baseline still said fail_to_run. I updated that expected-accuracy row to pass and resubmitted the PR.

**Review by eellison — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?
